### PR TITLE
Implement logaddexp

### LINKED ISSUE #38377: Implementing logaddexp, logaddexp2
I'd like implement logaddexp for cpu and cuda mentioned in #38349. these functions are perticularly useful when doing long probability accumulation and maintaining good numerical precesion. 

cc @mruberry

### PR DESCRIPTION
**Pull request by cloudhan**
Resolve #38377
Related #38349 

This op should be disambiguated with `logsumexp` which do a reduction on a tensor over a specific axis.

### CODE REVIEW

**Comment by ezyang** [comment]
Remove WIP from the title when you're ready for review

**Comment by cloudhan** [comment]
@mruberry anything left?

**Comment by mruberry** [comment]
> @mruberry anything left?

Just mark it as no longer WIP/draft when you're ready for review.

File: aten/src/ATen/native/BinaryOps.cpp [hunk 1]
```diff
@@ -742,6 +744,28 @@ Tensor& fmod_(Tensor& self, Scalar other) {
   return at::fmod_out(self, self, other);
 }
 
+Tensor& logaddexp_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  auto iter = TensorIterator::binary_op(result, self, other);
```
Comment (mruberry) [inline aten/src/ATen/native/BinaryOps.cpp:748]: Do you need a `/*check_mem_overlap=*/=true` here? What if result is self or other?

File: aten/src/ATen/native/BinaryOps.cpp [hunk 2]
```diff
@@ -742,6 +744,28 @@ Tensor& fmod_(Tensor& self, Scalar other) {
...
+  return result;
+}
+
+Tensor logaddexp(const Tensor& self, const Tensor& other) {
+  Tensor result = at::empty({0}, self.options());
+  return at::logaddexp_out(result, self, other);
+}
+
+Tensor& logaddexp2_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  auto iter = TensorIterator::binary_op(result, self, other);
```
Comment (mruberry) [inline aten/src/ATen/native/BinaryOps.cpp:759]: Same `check_mem_overlap` question as above.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 3]
```diff
@@ -595,6 +595,39 @@ void fmod_scalar_kernel(TensorIterator& iter, Scalar divisor) {
 
 }
 
+void logaddexp_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "logaddexp_cpu", [&]() {
+    cpu_kernel_vec(
+        iter,
+        [=](scalar_t a, scalar_t b) -> scalar_t {
+          scalar_t m = std::max(a, b);
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:603]: What happens if a == b == inf or -inf?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 4]
```diff
@@ -595,6 +595,39 @@ void fmod_scalar_kernel(TensorIterator& iter, Scalar divisor) {
...
+        });
+  });
+}
+
+void logaddexp2_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "logaddexp2_cpu", [&]() {
+    cpu_kernel_vec(
+        iter,
+        [=](scalar_t a, scalar_t b) -> scalar_t {
+          scalar_t m = std::max(a, b);
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:619]: Same question when a == b == +/- inf

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -514,6 +514,14 @@
 - name: log2(Tensor self) -> Tensor
   self: grad / (self * 0.6931471805599453)
 
+- name: logaddexp(Tensor self, Tensor other) -> Tensor
```
Comment (mruberry) [inline tools/autograd/derivatives.yaml:517]: @albanD Take a look, would you?

File: test/test_torch.py [hunk 6]
```diff
@@ -313,6 +313,38 @@ def test_dim_reduction_less_than_64(self):
         with self.assertRaisesRegex(RuntimeError, "PyTorch doesn't support reduction operations for dim>=64"):
             torch.sum(x, -1)
 
+    def _test_logaddexp(self, base2=False):
```
Comment (mruberry) [inline test/test_torch.py:316]: @albanD What's our plan for testing the derivatives of new functions like this?

File: test/test_torch.py [hunk 7]
```diff
@@ -313,6 +313,38 @@ def test_dim_reduction_less_than_64(self):
...
+        import numpy as np
```
Comment (mruberry) [inline test/test_torch.py:317]: Don't import NumPy, instead assert `TEST_NUMPY`

File: test/test_torch.py [hunk 8]
```diff
@@ -313,6 +313,38 @@ def test_dim_reduction_less_than_64(self):
...
+        self.assertTrue(np.allclose(ours.numpy(), gt))
+
+        # numerical stability
+        a *= 10000
+        b *= 10000
+        gt = gt_func(a.numpy(), b.numpy())
+        ours = our_func(a, b)
+        self.assertTrue(np.allclose(ours.numpy(), gt))
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
```
Comment (mruberry) [inline test/test_torch.py:340]: You should move these tests into TestTorchDeviceType so you can run them on both the CPU and GPU. Right now you're just testing the CPU. Then see the helper function "_np_compare," which you can use to simplify your tests, and the @dtypes decorator so you can test multiple dtypes. Since you're only enabling logaddexp on float types, you should test at least torch.long (assert throws RuntimeError), torch.float32, and torch.complex64 (assert throws RuntimeError). That way if someone implements logaddexp for complex64, for example, they'll know to update this test. 

_np_compare works by taking values to test at. Your value generation is OK but you should add some more "interesting" values like -math.pi, 0, and math.pi, as well as extremal values like nan, -inf, and inf.

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -2951,6 +2951,51 @@ def merge_dicts(*dicts):
...
+in statistics where the calculated probabilities of events may be so small as to
+exceed the range of normal floating point numbers. In such cases the logarithm
+of the calculated probability is stored. This function allows adding
+probabilities stored in such a fashion.
+
+This op should be disambiguated with :func:`torch.logsumexp` which do a
+reduction with a tensor over a specific axis.
+
+Args:
+    {input}
```
Comment (mruberry) [inline torch/_torch_docs.py:2970]: "- the input tensor" is what torch.add says here. It's not the most exciting descriptor but let's be consistent.

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -2951,6 +2951,51 @@ def merge_dicts(*dicts):
...
+    other (Tensor): the tensor to compute logaddexp with
```
Comment (mruberry) [inline torch/_torch_docs.py:2971]: This comment is a little misleading. Maybe something like: "the tensor whose exponential is added to the exponential of input before the log is taken"?

File: torch/_torch_docs.py [hunk 11]
```diff
@@ -2951,6 +2951,51 @@ def merge_dicts(*dicts):
...
+    {out}
```
Comment (mruberry) [inline torch/_torch_docs.py:2972]: out is an (optional) keyword argument. See what the torch.add docs do.

File: torch/_torch_docs.py [hunk 12]
```diff
@@ -2951,6 +2951,51 @@ def merge_dicts(*dicts):
...
+
+Example::
+
+    >>> torch.logaddexp(torch.tensor([-1.0]), torch.tensor([-1.0, -2, -3]))
```
Comment (mruberry) [inline torch/_torch_docs.py:2976]: These examples are helpful but I think you only need a single example showing the addition of multiple small probabilities, like the NumPy documentation, since that's the scenario that logaddexp is designed for.

File: torch/_torch_docs.py [hunk 13]
```diff
@@ -2951,6 +2951,51 @@ def merge_dicts(*dicts):
...
+
+Logarithm of the sum of exponentiations of the inputs in base-2.
+
+Calculates pointwise :math:`\log_2\left(2^x + 2^y\right)`. See
+:func:`torch.logaddexp` for more details.
+
+Args:
+    {input}
+    other (Tensor): the tensor to compute logaddexp2 with
+    {out}
```
Comment (mruberry) [inline torch/_torch_docs.py:2996]: Same language/doc changes here as with the above.

**Review by mruberry — CHANGES_REQUESTED** [review]
This PR is looking really good. I requested changes to the tests and docs and have a couple additional questions. I also want to check with @albanD if/how we're planning to validate the gradients of these functions.

Once we get that fixed up I think this PR will be good to go!

File: test/test_torch.py (hunk 6 above)
Comment (cloudhan) [reply test/test_torch.py:316]: I am also curious.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 1 above)
Comment (cloudhan) [reply aten/src/ATen/native/BinaryOps.cpp:748]: there will be no problem if result is self or other, but problem might occur when view of tensor overlap with other view, so `check_mem_overlap` is definitely needed here.

File: torch/_torch_docs.py (hunk 12 above)
Comment (cloudhan) [reply torch/_torch_docs.py:2976]: it is intended, just to show the numerical stability

File: torch/_torch_docs.py (hunk 9 above)
Comment (cloudhan) [reply torch/_torch_docs.py:2970]: they are automatically generated.

File: torch/_torch_docs.py (hunk 10 above)
Comment (cloudhan) [reply torch/_torch_docs.py:2971]: your suggestion is too verbose, may be I should just delete it and let the generator generate `the second input tensor` for it?

Comment (mruberry) [reply torch/_torch_docs.py:2971]: Sure.

File: tools/autograd/derivatives.yaml (hunk 5 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:517]: The formula looks good.
For testing, you want to add an entry [here](https://github.com/pytorch/pytorch/blob/378956b48143d5858cc183c885d39a7aa64182c9/torch/testing/_internal/common_methods_invocations.py#L563) to make sure the gradients will be properly checked.

File: tools/autograd/derivatives.yaml [hunk 14]
```diff
@@ -514,6 +514,14 @@
...
+  self: grad / (1 + exp(other - self))
+  other: grad / (1 + exp(self - other))
+
+- name: logaddexp2(Tensor self, Tensor other) -> Tensor
+  self: grad / (1 + pow(2, other - self))
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:522]: Isn't this one missing a `* 1/ln(2)` for the derivative of `log_2()`?

Comment (cloudhan) [reply tools/autograd/derivatives.yaml:522]: not exactly, log2(x) = ln(x)/ln(2), so ∂/∂x log2(x) = 1/(x ln(2)), but for  ∂/∂x 2^x, your got another ln(2) in the numerator, so they cancel out. When in doubt, consult the [Wolfram Alpha](https://www.wolframalpha.com/input/?i=log2%282%5Ex+%2B+2%5Ey%29) ;)

Comment (albanD) [reply tools/autograd/derivatives.yaml:522]: Ho right!
I was expecting something like that because the gradcheck is passing :D

**Review by albanD — APPROVED** [review]
LGTM
I'll let @mruberry  do a final pass.

File: test/test_torch.py [hunk 15]
```diff
@@ -9581,6 +9581,44 @@ def test_lerp(self, device):
                 expected = start + weight * (end - start)
                 self.assertEqual(expected, actual)
 
+    def _test_logaddexp(self, device, dtype, base2=False):
```
Comment (mruberry) [inline test/test_torch.py:9584]: Nit: don't include a default value for base2 (default values are appropriate for user APIs, but this is a developer API). If you want to improve readability you can also make base2 a kwarg-only argument:

```
_test_logaddexp(self, device, dtype, *, base2):
```

File: test/test_torch.py [hunk 16]
```diff
@@ -9581,6 +9581,44 @@ def test_lerp(self, device):
...
+        if base2:
+            gt_func = np.logaddexp2
+            our_func = torch.logaddexp2
+        else:
+            gt_func = np.logaddexp
+            our_func = torch.logaddexp
+
+        a = torch.randn(51, 2, dtype=dtype, device=device)
```
Comment (mruberry) [inline test/test_torch.py:9592]: This is good but to test vectorized and non-vectorized code paths you want to test with one array of at least 64 contiguous elements and another of fewer than four elements:

```
def _helper(vals_a, vals_b):
  a = torch.tensor(vals_a, device=device, dtype=dtype)
  b = torch.tensor(vals_b, device=device, dtype=dtype)
  
  actual = fn(a, b)
  expected = torch.from_numpy(ref(a.cpu().numpy(), b.cpu().numpy()))
  
  self.assertEqual(actual, expected)

vals_a = list((random.random() - .5 for _ in range(64)))
vals_b = list((random.random() - .5 for _ in range(64)))
_helper(vals_a, vals_b)
_helper(vals_a[:4], vals_b[:4])

vals_a = list(((random.random() - .5) * 1000 for _ in range(64)))
vals_b = list(((random.random() - .5) * 1000 for _ in range(64)))
_helper(vals_a, vals_b)
_helper(vals_a[:4], vals_b[:4])

# Tests nonfinite values
vals_a = [float('inf'), float('-inf'), float('inf'), float('nan')]
vals_b = [(float('inf'), float('-inf'), float('-inf'), float('nan')]
_helper(vals_a, vals_b)
```

File: test/test_torch.py [hunk 17]
```diff
@@ -9581,6 +9581,44 @@ def test_lerp(self, device):
...
+        b = torch.randn(51, 2, dtype=dtype, device=device)
+        gt = torch.tensor(gt_func(a.cpu().numpy(), b.cpu().numpy()), device=device)
+        ours = our_func(a, b)
+        self.assertEqual(ours.shape, gt.shape)
+        self.assertTrue(torch.allclose(ours, gt))
```
Comment (mruberry) [inline test/test_torch.py:9597]: See the sample body I posted. You can just use self.assertEqual for comparison

File: test/test_torch.py [hunk 18]
```diff
@@ -9581,6 +9581,44 @@ def test_lerp(self, device):
...
+
+        # numerical stability
+        a *= 10000
+        b *= 10000
+        a[0] = inf
```
Comment (mruberry) [inline test/test_torch.py:9602]: It's really great how thorough you're being here. The sample code should simplify this further.

File: torch/_torch_docs.py [hunk 19]
```diff
@@ -2984,6 +2984,55 @@ def merge_dicts(*dicts):
...
+
+Logarithm of the sum of exponentiations of the inputs.
+
+Calculates pointwise :math:`\log\left(e^x + e^y\right)`. This function is useful
+in statistics where the calculated probabilities of events may be so small as to
+exceed the range of normal floating point numbers. In such cases the logarithm
+of the calculated probability is stored. This function allows adding
+probabilities stored in such a fashion.
+
+This op should be disambiguated with :func:`torch.logsumexp` which do a
```
Comment (mruberry) [inline torch/_torch_docs.py:2999]: "which do a reduction with a tensor over a specific axis" -> "which performs a reduction on a single tensor"

File: torch/_torch_docs.py (hunk 9 above)
Comment (mruberry) [reply torch/_torch_docs.py:3003]: Right. Thanks for reminding me.

**Review by mruberry — CHANGES_REQUESTED** [review]
This is really great. Just a few minor doc comments and a suggestion for how the test can be simplified and we can ensure coverage of vectorized and non-vectorized. Writing tests can be, ironically, the hardest part of a PR and the least fun, but I gave you a sample which should do most of the work. I really appreciate how thorough your current test is, it just needs a small adjustment.

**Comment by cloudhan** [comment]
@mruberry

**Review by mruberry — APPROVED** [review]
Nice work, @cloudhan!

If you're still interested in contribution functions to PyTorch I encourage you to pick one of the more challenging functions in https://github.com/pytorch/pytorch/issues/38349, like count_nonzero or divmod.
